### PR TITLE
Modelclass

### PR DESCRIPTION
**Pull request by Priya2698**
Implementation for Model class to wrap Pytorch and Huggingface Models, obtaining model outputs, activations, and evaluation metrics.

### CODE REVIEW

File: meerkat/model/metrics.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    elif metric == "dice_score":
+        # Calculate Dice Score
+        score = dice(predictions=predictions, labels=labels)
+    else:
+        raise NotImplementedError
+
+    return score
+
+
+def format_summary(x: str) -> str:
```
Comment (krandiash) [inline meerkat/model/metrics.py:181]: Does this get used anywhere? If not, I'd throw it out

File: meerkat/model/model.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        if not device:
+            self.device = "cpu"
+            if torch.cuda.is_available():
+                self.device = "cuda:0"
+
+    def to(self, device: str):
+        self.device = device
+        return self.model.to(device)
+
+    """
```
Comment (seyuboglu) [inline meerkat/model/model.py:47]: `__call__ ` in `nn.Module` typically just calls forward. If we subclass `nn.Module` we can just remove this commented out block

File: meerkat/model/tensormodel.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+from meerkat.columns.tensor_column import TensorColumn
+from meerkat.model.activation import ActivationOp
+from meerkat.model.model import Model
+
+
+class TensorModel(Model):
+    def __init__(
+        self,
+        # identifier: str,
+        model,
```
Comment (seyuboglu) [inline meerkat/model/tensormodel.py:24]: Add a type hint for `nn.Module`

File: meerkat/model/tensormodel.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+            input_batch = batch[input_columns[0]].data.to(self.device)
+
+        return input_batch
+
+    def predict_batch(
+        self, dataset: DataPanel, input_columns: List[str], batch_size=32
+    ):
+        # Gathers predictions for entire dataset
+
+        predictions = []
```
Comment (seyuboglu) [inline meerkat/model/tensormodel.py:84]: This can be done with an update and can be moved up to the `model` base class since the implementation is shared with huggingfacemodel

File: meerkat/model/model.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
+from typing import Dict, List
+
+import torch
+
+from meerkat import DataPanel
+from meerkat.columns.prediction_column import ClassificationOutputColumn
+from meerkat.model.metrics import compute_metric
+
+
```
Comment (seyuboglu) [inline meerkat/model/model.py:9]: Something to consider for a future PR: I think that a lot of the functionality implemented in the `Model` class and it's subclasses (`HuggingFaceModel`, `TensorModel`) could be implemented in a functional API. We can discuss this for a future PR, but this looks good for this PR.

**Review by seyuboglu — CHANGES_REQUESTED** [review]
Left a few comments, but otherwise lgtm!

File: meerkat/columns/instances_column.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+            total=(len(self) // batch_size + int(len(self) % batch_size != 0)),
+        ):
+            batch_data = [instance.get_fields()[field] for instance in batch]
+            data = list(itertools.chain(data, batch_data))
+
+        data_col = ListColumn(data)
+
+        return data_col
+
+    def get_nms(
```
Comment (ad12) [inline meerkat/columns/instances_column.py:51]: It could be useful if NMS should return a new `InstancesColumn` where the boxes that have been suppressed are removed from each instance. The indices may not mean much at the `InstancesColumn` level b/c it'd be hard for the user to make sense of indices that aren't completely visible to them

File: meerkat/columns/segmentation_column.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+from meerkat.columns.tensor_column import TensorColumn
+from meerkat.datapanel import DataPanel
+from meerkat.tools.lazy_loader import LazyLoader
+
+cv2 = LazyLoader("cv2")
+
+Columnable = Union[Sequence, np.ndarray, pd.Series, torch.Tensor]
+
+
+def _convert_rle2mask(
```
Comment (ad12) [inline meerkat/columns/segmentation_column.py:24]: nit: move private methods to end of the file

File: meerkat/columns/text_column.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
+from __future__ import annotations
+
+from typing import Sequence
+
+from meerkat.columns.list_column import ListColumn
+
+
+class TextOutputColumn(ListColumn):
```
Comment (ad12) [inline meerkat/columns/text_column.py:8]: Is there any functionality that is different from a `ListColumn`? Maybe we can punt this for now?

File: meerkat/model/metrics.py [hunk 9]
```diff
@@ -0,0 +1,184 @@
+import re
```
Comment (ad12) [inline meerkat/model/metrics.py:1]: I think we should use `torchmetrics` here. It helps with reproducibility and takes the work out of distributed metrics in the future

File: meerkat/model/tensormodel.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+from meerkat.columns.embedding_column import EmbeddingColumn
+from meerkat.columns.instances_column import InstancesColumn
+from meerkat.columns.prediction_column import ClassificationOutputColumn
+from meerkat.columns.segmentation_column import SegmentationOutputColumn
+from meerkat.columns.tensor_column import TensorColumn
+from meerkat.model.activation import ActivationOp
+from meerkat.model.model import Model
+
+
+class TensorModel(Model):
```
Comment (ad12) [inline meerkat/model/tensormodel.py:20]: I think defining different "kinds" of outputs (classification, time series, etc.) may become a bit unwieldy in this class. It's useful to have those utilities, but in the class definition may not be the right place. Users may have different ways of getting their outputs, so as Sabri mentioned, it may be good to make these methods functions instead.

**Review by ad12 — CHANGES_REQUESTED** [review]

File: meerkat/model/activation.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+    def __init__(self):
+        self.activation = None
+
+    def add_hook(self, module, input, output):
+        self.activation = output
+
+
+# TODO: test on an NLP model
+class ActivationOp:
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply meerkat/model/activation.py:29]: For extracting activations in tensormodel.py and huggingfacemodel.py

File: meerkat/columns/text_column.py (hunk 8 above)
Comment (Priya2698) [reply meerkat/columns/text_column.py:8]: Yes, for now this is redundant. I'll remove this

File: meerkat/columns/instances_column.py [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+
+class InstancesColumn(ListColumn):
+    def __init__(self, data: Sequence = None, *args, **kwargs):
+
+        super(InstancesColumn, self).__init__(data=data, *args, **kwargs)
+
+    def num_instances(
+        self, batch_size: int = 32
+    ) -> ListColumn:  # TODO(Priya): ListColumn or some other type
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply meerkat/columns/instances_column.py:21]: TensorColumn or NumpyArrayColumn?

File: meerkat/model/metrics.py (hunk 9 above)
Comment (Priya2698) [reply meerkat/model/metrics.py:1]: I have switched to torchmetrics from sklearn for accuracy and f1 scores. IoU and Dice score already use torchmetrics

File: meerkat/__init__.py [hunk 13]
```diff
@@ -17,16 +17,20 @@
...
 from meerkat.columns.pandas_column import PandasSeriesColumn
 from meerkat.columns.prediction_column import ClassificationOutputColumn
+from meerkat.columns.segmentation_column import SegmentationOutputColumn
 from meerkat.columns.spacy_column import SpacyColumn
 from meerkat.columns.tensor_column import TensorColumn
 from meerkat.columns.video_column import VideoColumn
 from meerkat.columns.volume_column import MedicalVolumeColumn
 from meerkat.datapanel import DataPanel
+from meerkat.model.huggingfacemodel import HuggingfaceModel
+from meerkat.model.tensormodel import TensorModel
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply meerkat/__init__.py]: Moved all this and ClassificationOutputColumn to nn submodule. Let me know if that looks okay now

File: meerkat/columns/instances_column.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Priya2698) [reply meerkat/columns/instances_column.py]: Switched to TensorColumn

File: meerkat/model/huggingfacemodel.py [hunk 14]
```diff
@@ -0,0 +1,258 @@
...
+
+        # Return the converted batch
+        return input_batch
+
+    def predict_batch(
+        self, dataset: DataPanel, input_columns: List[str], batch_size=32
+    ):
+        # Gathers predictions for entire dataset
+
+        predictions = []
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply meerkat/model/huggingfacemodel.py]: Changed to map for now. Will switch to update once I can specify output_type there to remove some redundant code

File: meerkat/model/model.py [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+# TODO(Priya): Move some general functions here
+class Model:
```
[the remark this answers is not in the document]
Comment (Priya2698) [reply meerkat/model/model.py]: Added this

File: meerkat/model/tensormodel.py (hunk 4 above)
Comment (Priya2698) [reply meerkat/model/tensormodel.py]: A common _predict method is now present in `Model` class

File: meerkat/columns/instances_column.py (hunk 6 above)
Comment (Priya2698) [reply meerkat/columns/instances_column.py]: I am returning a ListColumn of boxes tensors now

File: meerkat/__init__.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (seyuboglu) [reply meerkat/__init__.py]: Looks awesome

**Review by seyuboglu — APPROVED** [review]

**Comment by seyuboglu** [comment]
lgtm!

**Review by ad12 — APPROVED** [review]
lgtm!
